src/include/network/notifiable_task.h

```diff
@@ -0,0 +1,244 @@
+//===----------------------------------------------------------------------===//
+//
+//                         Peloton
+//
+// notifiable_task.h
+//
+// Identification: src/include/network/notifiable_task.h
+//
+// Copyright (c) 2015-17, Carnegie Mellon University Database Group
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <csignal>
+#include <cstdio>
+#include <cstdlib>
+#include <cstring>
+#include <vector>
+#include <unordered_set>
+
+#include <sys/file.h>
+#include <event2/thread.h>
+
+#include "common/exception.h"
+#include "common/logger.h"
+#include "container/lock_free_queue.h"
+#include "network_state.h"
+#include "error_util.h"
+
+namespace peloton {
+namespace network {
+
+/**
+ * Convenient MACRO to use a method as a libevent callback function. Example
+ * usage:
+ *
+ * (..., METHOD_AS_CALLBACK(ConnectionDispatcherTask, DispatchConnection), obj)
+ *
+ * Would call DispatchConnection method on ConnectionDispatcherTask. obj must be
+ * an instance
+ * of ConnectionDispatcherTask. The method being invoked must have signature
+ * void(int, short),
+ * where int is the fd and short is the flags supplied by libevent.
+ *
+ */
+#define METHOD_AS_CALLBACK(type, method)         \
+  [](int fd, short flags, void *arg) {           \
+    static_cast<type *>(arg)->method(fd, flags); \
+  }
+
+/**
+ * @brief NotifiableTasks can be configured to handle events with callbacks, and
+ * executes within an event loop
+ *
+ * More specifically, NotifiableTasks are backed by libevent, and takes care of
+ * memory management with the library.
+ */
+class NotifiableTask {
+ public:
+  /**
+   * Constructs a new NotifiableTask instance.
+   * @param task_id a unique id assigned to this task
+   */
+  explicit NotifiableTask(int task_id) : task_id_(task_id) {
+    base_ = EventUtil::EventBaseNew();
+    // TODO(tianyu) Determine whether we actually need this line. Tianyi says we
+    // need it, libevent documentation says no
```

Comment:
Which part of libevent book are you refering to? I am actually not sure.

Reply:
the doxygen doc provided in event.h